### ISSUE TITLE
Support for external frameworks (CDI/Weld, TestFX and JMockit in this case)

### ISSUE THREAD

**Issue by adam-waldenberg** [jqwik-team/jqwik#386]
We recently wrote some jqwik hooks for our test suite in https://github.com/unigrid-project/janus-java. So far, we got plugins for TestFX and Weld. I thought that it might be useful to others. There is also support for JMockit.

The code can be found here:
https://github.com/unigrid-project/janus-java/tree/master/fx/src/test/java/org/unigrid/janus/jqwik

For CDI/Weld, it allows for the following in a jqwik test:

```java
@WeldSetup(Address.class)
public class AddressTest extends BaseMockedWeldTest {
	@Inject
	private Address address;

	@Example
	public boolean testAddressProperties() {
		final String addr = "HAjsFi8JShq9Hfx5xYseGMoy8Mzbbo3Reu";

		address.setAddress(addr);
		return addr.endsWith(address.getAddress());
	}
}
```

And using TestFX and JMockit:

```java
@FxResource(clazz = MainWindow.class, name = "mainWindow.fxml")
public class AddressControllerTest extends BaseFxTest {
	@Inject
	private FxRobot robot;

	@Inject
	private RPCService rpc;

	@FxStart
	public void startWithStage(Stage stage) {
		System.out.println("I'm starting a test run and I am only here for demo purposes!");
	}

	@FxStop
	public void stop() {
		System.out.println("I'm stopping a test run and I am only here for demo purposes!");
	}

	@BeforeContainer
	private static void before() {
		new ResponseMockUp();
		new JerseyInvocationMockUp();
		new WebTargetMockUp();
		new DaemonMockUp();
	}

	@Example
	public void shouldHideZeroBalances() {
		Eager.instantiate(rpc);
		waitForScene();

		robot.clickOn("#btnAddress");
		robot.clickOn("#chkAddress");
		verifyThat("#tblAddresses", hasTableCell(2000.0));
	}
}
```

I saw there is a https://github.com/jlink/jqwik-spring project for Spring support. Question I have is if there is any interest in integrating Weld/FX/Jmockit support into jqwik. Maybe an idea would be to have additional build modules inside jqwik  (separate artifacts for each module)?

**Comment by jlink** [comment]
@adam-waldenberg Thanks for suggesting that. Currently, there are already three more or less active jqwik extension modules: 
https://github.com/jlink/jqwik-spring
https://github.com/jlink/jqwik-testcontainers
https://github.com/Befrish/jqwik-vavr

All of them are decoupled (by version and repository) from jqwik's main repo. I consider that a good thing, b/c extension's maintainance cycle is usually different that that of jqwik itself.

Having three more extensions is perfectly doable, so I'm will to set that up.

There're probably good arguments for collecting all extensions in a single repo - and maybe even for releasing them together. I just don't have the free time to do that work myself, so if you'd like to put everything in one place, I'd be willing to set up the repo and the parts necessary for deployment and releasing.

What's definitely missing is a section on the website / user guide to point to all jqwik extensions. I guess there's already a GitHub issue for that.

**Comment by adam-waldenberg** [comment]
@jlink Awesome.

Yeah.. Hmmm.. I wonder what the best way to approach it would be. It should definitely be separated as all these will be maintained independently. Shoving them in the same repo probably wont work well in the long run unless someone is willing to be on top of these constantly ... If they grow in number thats going to be an endeavor in its own. At the same time, it would be nice to get everything in one place - especially when checking out and building jqwik.

Maybe git submodules is the way to go. Then it can be separated from the main repo but still be accessible from one place.

**Comment by jlink-workshop** [comment]
How do git submodules work?

The alternative is to create a jqwik organization and collect all relevant repos and projects there.

**Comment by adam-waldenberg** [comment]
Basically it's just a directory with a link to another git repository ... Then you can set up your build cycle to include them if they are present (when you git pull, you can actually choose to omit them, which is the default behavior).

**Comment by jlink** [comment]
@adam-waldenberg If I set up a common repo or an org for jqwik extensions, would you be willing to add and maintain the extensions for CDI/Weld, TestFX and JMockit?

**Comment by adam-waldenberg** [comment]
@jlink Yes I would be happy to do that.

**Comment by adam-waldenberg** [comment]
@jlink Small update, we are currently working on Arquillian integration now as well and making some good progress.  So that is going to be another one for this list.

**Comment by jlink** [comment]
@adam-waldenberg I just created an org for jqwik-related development: https://github.com/jqwik-team/
Just tell me, which repos I should create and add you as the maintainer.

**Comment by adam-waldenberg** [comment]
@jlink Awesome. As soon as I have time I will clean it up ... and move it in. I think we need four of them;

1. Weld
2. TestFX
3. JMockit
4. Arquillian (still working on this one)

**Comment by jlink** [comment]
@adam-waldenberg I’ll call the repos jqwik-weld, jqwik-testfx, jqwik-jmockit and jqwik-arquillian. I‘ll invite you as contributor/maintainer so that you can set them up. If fine for you, I’ll do it next week since I’m on vacation and only have a phone for internet access.

**Comment by adam-waldenberg** [comment]
> @adam-waldenberg I’ll call the repos jqwik-weld, jqwik-testfx, jqwik-jmockit and jqwik-arquillian. I‘ll invite you as contributor/maintainer so that you can set them up. If fine for you, I’ll do it next week since I’m on vacation and only have a phone for internet access.

Sounds good.

**Comment by jlink** [comment]
Done.

See https://github.com/jqwik-team/jqwik-testfx, https://github.com/jqwik-team/jqwik-jmockit, https://github.com/jqwik-team/jqwik-weld, https://github.com/jqwik-team/jqwik-arquillian

**Comment by adam-waldenberg** [comment]
@jlink I know its a long time after... But I completed an implementation for Arquillian ...  I don't have push access to that repo though.... Do you want to enable that or do you prefer it as a pull request in there ?

**Comment by jlink** [comment]
Give me a few days to figure it out. Or just make a PR.

**Comment by adam-waldenberg** [comment]
No rush :)

**Comment by jlink** [comment]
@adam-waldenberg I gave you write access to the repository. Just give me a heads-up when you think that the Arquillian extension is ready for a first publishing.

Mind that there is a no-AI contribution policy for all jqwik-related projects. But you've probably noticed that already.

**Comment by adam-waldenberg** [comment]
Yeah I think that is a good call... While I do use AI for many tasks, I personally don't trust it to completely write my code. The times I tried it it made a complete mess most of the time. The problem is that the training data the AI's use is mostly badly structured code and I don't think that will change any time soon..

**Comment by adam-waldenberg** [comment]
Pushed and published. 'create-release.sh' comes from another project and I mostly modified it for the arquillian module. Hopefully it kind of works.... I also added Lombok into the build.... Not sure if we have a convention but if you think it should be removed it was added in a single commit so it can be easily reversed.

This was a big pain in the ass to get working .... and I kind of had something already a year ago - but didn't have time to get it up here.... This verson cleans that up and makes it presentable to everyone. I had to do some guesses on the Arquillian side and some workaround on the JQwik side to make it work properly and clean up between runs... But I think I got it right.

I also have something half-done for JMockit, but that one still has a few problems and I will see when I can get it sorted to a point where I feel it can be shared. It works great in the projects I use it but is very limited to those specific parts of JMockit that I am using there... For it to be presentable it needs to support the full suite. Same goes for Weld - not everything works there yet either.

The TestFX repo I think we can drop - unless somebody else wants to maintain it.
